### ISSUE TITLE
Allow the admin to request existing users to reaccept the new Terms & Conditions

### ISSUE THREAD

**Issue by cgeorgilakis** [keycloak/keycloak#26124]
### Description

In our production environment of Keycloak, whenever the "Terms and Conditions" change, we want to ask all users already registered to keycloak to re-approve the new terms and conditions.
Currently, keycloak does not allow to reset the “terms_and_conditions” user_required_action flag for all existing users.
We propose to add an extra kebab button on the "Terms and Conditions", to support resetting this action for all existing users. 
This might also be useful for the other Required actions, such as the “Update password” in cases that an administrator would like to ask all existing users to update their passwords.

### Details

[Old jira issue](https://issues.redhat.com/browse/KEYCLOAK-17379) exists for it.
Based on discussion in [related jira issue](https://github.com/keycloak/keycloak/issues/19275).

**Comment by jonkoops** [comment]
The way I am imagining this for individual users, is that we make this something that we can add to the 'Required User Actions', so that the user can be forced to re-accept the Terms and Conditions:

![A screenshot showing "Accept Terms and Conditions" as an option for required user actions.](https://github.com/keycloak/keycloak/assets/695720/139a4f31-86da-4597-b81d-f5a2ed735397)

This would then work exactly the same as updating the password, or other similar actions. From there, we can start thinking about how this would be assigned on a batch of users under certain conditions. Perhaps we need to have a 'Terms and Conditions' tab somewhere in the realm settings to better manage these settings.

cc @xianli123 @ssilvert

**Comment by ssilvert** [comment]
> The way I am imagining this for individual users, is that we make this something that we can add to the 'Required User Actions', so that the user can be forced to re-accept the Terms and Conditions:
> 
> ![A screenshot showing "Accept Terms and Conditions" as an option for required user actions.](https://private-user-images.githubusercontent.com/695720/295911359-139a4f31-86da-4597-b81d-f5a2ed735397.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUwMTEwNjgsIm5iZiI6MTcwNTAxMDc2OCwicGF0aCI6Ii82OTU3MjAvMjk1OTExMzU5LTEzOWE0ZjMxLTg2ZGEtNDU5Ny1iODFkLWY1YTJlZDczNTM5Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDExMVQyMjA2MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kODQ0ODI0NmUxYjI0MDQ2MjhmNzI5YmFkYjg3MjEwMWQzYmRiYjgxNTc3ZDE3ZGMzOThkZDBjMjYzZjJhYmY4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.JaPGoIEk14q2u-6Rb4aNWY4PmaXG1nLOgs9aG3gm6fU)
> 
> This would then work exactly the same as updating the password, or other similar actions. From there, we can start thinking about how this would be assigned on a batch of users under certain conditions. Perhaps we need to have a 'Terms and Conditions' tab somewhere in the realm settings to better manage these settings.
> 
> cc @xianli123 @ssilvert

We already have this for a single user, like you showed in your screen shot.

If the core team give us an API for a batch of users then we can just allow the admin to call it from the Authentication --> Required Actions page.  I can see where this might be useful for other required actions as well.

**Comment by jonkoops** [comment]
> We already have this for a single user, like you showed in your screen shot.

We do? I created that screenshot by editing the code as an example. Where and how can this be done? I couldn't find anything in the documentation.

**Comment by xianli123** [comment]
Hi @jonkoops here is a previous conversation related to the terms and conditions. https://github.com/keycloak/keycloak/pull/7895 It would be a new page to manage the terms and conditions. Is it what you are expecting?

**Comment by jonkoops** [comment]
@ssilvert You are correct, this feature already exists, but is [not properly documented](https://www.keycloak.org/docs/latest/server_admin/#proc-requiring-tac-agreement-at-registration_server_administration_guide). The required action can be enabled by going to `Authentication` :arrow_right: `Required actions` and enabling `Terms and Conditions` in the overview:

![A screenshot showing 'Terms and Conditions' as enabled.](https://github.com/keycloak/keycloak/assets/695720/a004bdbd-97d7-4f04-9bee-2e8968c52eae)

I will consider this issue resolved, and log a separate issue to improve the documentation on required user actions.

**Comment by jonkoops** [comment]
Created #26153 to handle the documentation of this feature.

**Comment by ssilvert** [comment]
I'm reopening this issue as the feature @cgeorgilakis is asking for does not exist.  He wants a way to set the T&C required action for all existing users.  This will require a new server side API.  

I'm going to remove the UI team for now as the core team needs to decide if and how we move forward.

**Comment by jonkoops** [comment]
> This will require a new server side API.

I doubt we want to implement an API just for that, unless we decide to create an API that allows bulk actions to be performed on all users. But I will leave that up to the core team.

**Comment by afgarcia86** [comment]
This would be really helpful.

**Comment by thomas-wangler** [comment]
Are there any updates on this?

**Comment by bjozturk** [comment]
This is much needed. Regulations require notifying users whenever updates are made to the terms and conditions. Additionally, having the ability to prompt all existing users to re-accept the updated ToS would be useful. This would provide a timestamped record.

**Comment by danifr** [comment]
I agree, it would be great something like this implemented.

Or rather, have some mechanism that if T&C change, all users need to reaccept them during the next login.

This could be also used to force all users to reset their passwords after a breach or similar.

**Comment by ch-lepp** [comment]
Does anyone know a feasible workaround while we wait for a solution?

We currently work with an SQL, but that seems hardly appropriate...

```
INSERT INTO user_required_action (user_id, required_action)
SELECT u.id, 'TERMS_AND_CONDITIONS'
FROM user_entity u
WHERE u.realm_id='myrealm'
AND u.service_account_client_link IS NULL
AND NOT EXISTS (
	SELECT 1 FROM user_required_action a
	WHERE u.id=a.user_id 
	AND a.required_action = 'TERMS_AND_CONDITIONS'
); 
```

**Comment by huyenvu2101** [comment]
@ch-lepp  Here’s what I did: Run a script (python or whatever):  obtain an admin token, then fetch the user list via the Admin API. Iterate through the IDs; if user["requiredActions"] doesn't include TERMS_AND_CONDITIONS and user have not accept the TC, send a PUT request to update the user.
Let AI do it for u, logs everything to a file to verify if it work

**Comment by danifr** [comment]
yep, this the python code I use (with https://python-keycloak.readthedocs.io/en/latest/getting_started.html):
```
import os
from datetime import datetime
from keycloak import KeycloakAdmin

admin_user = "XXXX"
admin_pass = "YYYYYYYY"

server_url = "https://yourkeycloak.server.org/auth/"
realm_name = "your_realm"

admin = KeycloakAdmin(
            server_url=server_url,
            username=admin_user,
            password=admin_pass,
            realm_name=realm_name,
            user_realm_name="master")

TC_CUTOFF = datetime(2026, 1, 1).timestamp()

def require_tc_acceptance(user: dict) -> None:
    actions = user.get("requiredActions", [])
    if "TERMS_AND_CONDITIONS" not in actions:
        actions.append("TERMS_AND_CONDITIONS")
        admin.update_user(user_id=user["id"], payload={"requiredActions": actions})

for user in admin.get_users({}):
    username = user["username"]
    tc_attr = user.get("attributes", {}).get("terms_and_conditions")

    if not tc_attr:
        print(f"{username} never accepted the T&C — flagging")
        require_tc_acceptance(user)
    elif int(tc_attr[0]) < TC_CUTOFF:
        accepted_on = datetime.fromtimestamp(int(tc_attr[0])).strftime("%c")
        print(f"{username} must re-accept T&C (accepted on {accepted_on})")
        require_tc_acceptance(user)
    else:
        print(f"{username}: T&C up to date")
```

**Comment by Swasthik-S-Nettar** [comment]
We previously encountered a similar requirement where we needed to re-enable the Terms & Conditions for users who had already accepted them. To achieve this, we created a shell script that identifies all users who had previously accepted the T&C, retrieves their user IDs, and then re-applies the T&C as a required action for each of those users.

Below is the shell script that we had used.
```shell
# Note:
# $1 = Hostname (e.g., keycloak.example.com)
# $2 = Context path (e.g., auth)
# $3 = Realm name (e.g., myrealm)
# $4 = Admin username
# $5 = Admin password
#
# Example run:
# ./reenable_tc.sh keycloak.example.com auth myrealm admin password123

# Step 1: Get Admin Token
TOKEN=$(curl -k -s -X POST "https://$1/$2/realms/$3/protocol/openid-connect/token" \
  -H "Content-Type: application/x-www-form-urlencoded" \
  -d "username=$4" -d "password=$5" -d "grant_type=password" -d "client_id=admin-cli" | jq -r '.access_token')
 
# Step 2: Find ALL users who have ACTUALLY accepted T&C (have the attribute)
curl -k -s -X GET "https://$1/$2/admin/realms/$3/users?max=10000" \
  -H "Authorization: Bearer $TOKEN" -H "Content-Type: application/json" | \
  jq '[.[] | select(.attributes.terms_and_conditions != null) | {username: .username, id: .id, acceptedOn: .attributes.terms_and_conditions[0]}]' > users_who_accepted_tc.json
 
# Step 3: Display the users
cat users_who_accepted_tc.json | jq '.'
 
# Step 4: Re-enable T&C for each user who accepted it
USER_IDS=$(cat users_who_accepted_tc.json | jq -r '.[].id')
 
for USER_ID in $USER_IDS; do
  USERNAME=$(cat users_who_accepted_tc.json | jq -r ".[] | select(.id == \"$USER_ID\") | .username")
  echo "Re-enabling T&C for: $USERNAME ($USER_ID)"
  curl -k -s -X PUT "https://$1/$2/admin/realms/$3/users/$USER_ID" \
    -H "Authorization: Bearer $TOKEN" \
    -H "Content-Type: application/json" \
    -d '{"requiredActions": ["TERMS_AND_CONDITIONS"]}'
  echo "✓ T&C re-enabled for $USERNAME"
done
 
# Step 5: Verify all users now have T&C re-enabled
for USER_ID in $USER_IDS; do
  curl -k -s -X GET "https://$1/$2/admin/realms/$3/users/$USER_ID" \
    -H "Authorization: Bearer $TOKEN" | jq '{username: .username, requiredActions: .requiredActions, acceptedOn: .attributes.terms_and_conditions[0]}'
done

```
